File: Payload_Type/thanatos/mythic/command/sleep/sleep.go
package agentfunctions

import (
	"errors"
	"fmt"
	"strconv"
	"strings"

	agentstructs "github.com/MythicMeta/MythicContainer/agent_structs"
	"github.com/MythicMeta/MythicContainer/mythicrpc"
)

var sleepCommandDefinition = agentstructs.Command{
	Name:        "sleep",
	Description: "Change the agent's sleep interval. Suffix can either be [s, m, h]",
	HelpString:  "sleep [number][suffix] [jitter]",
	Version:     1,
	TaskCompletionFunctions: map[string]agentstructs.PTTaskCompletionFunction{
		"sleepPostRunActions": sleepPostRunActions,
	},

	MitreAttackMappings: []string{
		"T1029",
	},

	Author: "@M_alphaaa",
	CommandAttributes: agentstructs.CommandAttribute{
		SupportedOS: []string{
			agentstructs.SUPPORTED_OS_LINUX, agentstructs.SUPPORTED_OS_WINDOWS,
		},
		CommandIsBuiltin: true,
	},

	CommandParameters: []agentstructs.CommandParameter{
		{
			Name:             "interval",
			Description:      "Interval to sleep.",
			ModalDisplayName: "Interval to sleep.",
			ParameterType:    agentstructs.COMMAND_PARAMETER_TYPE_STRING,
			ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
				{
					ParameterIsRequired: true,
					UIModalPosition:     1,
				},
			},
		},

		{
			Name:             "jitter",
			Description:      "Sleep jitter.",
			ModalDisplayName: "Sleep jitter.",
			DefaultValue:     0,
			ParameterType:    agentstructs.COMMAND_PARAMETER_TYPE_NUMBER,
			ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
				{
					ParameterIsRequired: false,
					UIModalPosition:     2,
				},
			},
		},
	},

	TaskFunctionParseArgString: sleepParseArgsString,

	TaskFunctionParseArgDictionary: func(args *agentstructs.PTTaskMessageArgsData, input map[string]interface{}) error {
		args.LoadArgsFromDictionary(input)
		return nil
	},

	TaskFunctionCreateTasking: sleepCreateTasking,
}

func sleepParseArgsString(args *agentstructs.PTTaskMessageArgsData, input string) error {
	cmdLineArgs := strings.Split(input, " ")

	args.SetArgValue("interval", cmdLineArgs[0])

	if len(cmdLineArgs) > 1 {
		jitter, err := strconv.Atoi(cmdLineArgs[1])
		if err != nil {
			return err
		}

		args.SetArgValue("jitter", jitter)
	} else {
		args.SetArgValue("jitter", 0)
	}

	return nil
}

func sleepUpdateSleepInfo(callbackID int, output string) error {
	if _, err := mythicrpc.SendMythicRPCCallbackUpdate(mythicrpc.MythicRPCCallbackUpdateMessage{
		CallbackID: &callbackID,
		SleepInfo:  &output,
	}); err != nil {
		return errors.New(fmt.Sprintf("Failed to update sleep info: %s", err.Error()))
	}

	return nil
}

func sleepFormulateOutput(taskID int, output string) error {
	if _, err := mythicrpc.SendMythicRPCResponseCreate(mythicrpc.MythicRPCResponseCreateMessage{
		TaskID:   taskID,
		Response: []byte(output),
	}); err != nil {
		return errors.New(fmt.Sprintf("Failed to set the task output: %s", err.Error()))
	}

	return nil
}

func sleepPostRunActions(
	task *agentstructs.PTTaskMessageAllData,
	data *agentstructs.PTTaskMessageAllData,
	subtask *agentstructs.SubtaskGroupName,
) agentstructs.PTTaskCompletionFunctionMessageResponse {

	completionMessage := agentstructs.PTTaskCompletionFunctionMessageResponse{
		TaskID:       task.Task.ID,
		ParentTaskId: 0,
		Success:      false,
	}

	interval, err := task.Args.GetNumberArg("interval")
	if err != nil {
		completionMessage.Error = fmt.Sprintf("Failed to get sleep interval in post processing: %s", err.Error())
		return completionMessage
	}

	jitter, err := task.Args.GetNumberArg("jitter")
	if err != nil {
		completionMessage.Error = fmt.Sprintf("Failed to get sleep jitter in post processing: %s", err.Error())
		return completionMessage
	}

	var sleepInfo string
	if jitter == 0 {
		sleepInfo = fmt.Sprintf("Agent will checkin every %d second(s)", int(interval))
	} else {
		jitter /= 100
		startRange := interval - (float64(interval) * jitter)
		endRange := interval + (float64(interval) * jitter)

		sleepInfo = fmt.Sprintf("Agent will checkin between %d and %d second(s)", int(startRange), int(endRange))
	}

	if err := sleepFormulateOutput(task.Task.ID, sleepInfo); err != nil {
		completionMessage.Error = err.Error()
		return completionMessage
	}

	if err := sleepUpdateSleepInfo(task.Task.CallbackID, sleepInfo); err != nil {
		completionMessage.Error = err.Error()
		return completionMessage
	}

	completionMessage.Success = true
	return completionMessage
}

func sleepCreateTasking(task *agentstructs.PTTaskMessageAllData) agentstructs.PTTaskCreateTaskingMessageResponse {
	taskResponse := agentstructs.PTTaskCreateTaskingMessageResponse{
		Success: false,
		TaskID:  task.Task.ID,
	}

	intervalStr, err := task.Args.GetStringArg("interval")
	if err != nil {
		taskResponse.Error = fmt.Sprintf("Failed to get interval argument: %s", err.Error())
		return taskResponse
	}

	jitter, err := task.Args.GetNumberArg("jitter")
	if err != nil {
		taskResponse.Error = fmt.Sprintf("Failed to get jitter argument: %s", err.Error())
		return taskResponse
	}

	var timeUnit string
	conversionFactor := 1
	realInterval, err := strconv.Atoi(intervalStr)
	if err != nil {
		realInterval, err = strconv.Atoi(intervalStr[:len(intervalStr)-1])
		if err != nil {
			taskResponse.Error = fmt.Sprintf("Failed to parse the interval as an integer: %s", err.Error())
			return taskResponse
		}

		timeUnit = intervalStr[len(intervalStr)-1:]

		switch timeUnit {
		case "s":
			conversionFactor = 1
		case "m":
			conversionFactor = 60
		case "h":
			conversionFactor = 3600
		default:
			taskResponse.Error = fmt.Sprintf("Invalid interval suffix of '%s'. Must be either s, m, or h", timeUnit)
			return taskResponse
		}

		realInterval *= conversionFactor
	} else {
		timeUnit = "s"
	}

	if jitter < 0 {
		taskResponse.Error = "Jitter cannot be negative"
		return taskResponse
	}

	task.Args.RemoveArg("interval")
	task.Args.AddArg(agentstructs.CommandParameter{
		Name:          "interval",
		ParameterType: agentstructs.COMMAND_PARAMETER_TYPE_NUMBER,
	})

	task.Args.SetArgValue("interval", float64(realInterval))

	displayInterval := int(realInterval / conversionFactor)
	displayParams := fmt.Sprintf("interval = %d%s, jitter = %d%%", displayInterval, timeUnit, int(jitter))
	taskResponse.DisplayParams = &displayParams

	completionFunction := "sleepPostRunActions"
	taskResponse.CompletionFunctionName = &completionFunction

	taskResponse.Success = true
	return taskResponse
}

func init() {
	agentstructs.AllPayloadData.Get("thanatos").AddCommand(sleepCommandDefinition)
}

File: Payload_Type/thanatos/thanatos/mythic/agent_functions/sleep.py
import json
import traceback
from mythic_container.MythicCommandBase import (
    TaskArguments,
    CommandBase,
    CommandAttributes,
    SupportedOS,
    CommandParameter,
    ParameterGroupInfo,
    ParameterType,
    MythicTask,
    MythicStatus,
    PTTaskCompletionFunctionMessage,
    PTTaskCompletionFunctionMessageResponse,
)
from mythic_container.MythicGoRPC import (
    SendMythicRPCResponseCreate,
    MythicRPCResponseCreateMessage,
    SendMythicRPCCallbackUpdate,
    MythicRPCCallbackUpdateMessage,
)


class SleepArguments(TaskArguments):
    def __init__(self, command_line, **kwargs):
        super().__init__(command_line, **kwargs)
        self.args = [
            CommandParameter(
                name="interval",
                type=ParameterType.String,
                description="Interval to sleep.",
                display_name="Interval to sleep.",
                parameter_group_info=[ParameterGroupInfo(ui_position=1)],
            ),
            CommandParameter(
                name="jitter",
                type=ParameterType.Number,
                default_value=0,
                description="Sleep jitter.",
                display_name="Sleep jitter.",
                parameter_group_info=[ParameterGroupInfo(required=False, ui_position=2)],
            ),
        ]

    async def parse_arguments(self):
        if self.tasking_location == "command_line":
            if self.command_line[0] == "{":
                self.load_args_from_json_string(self.command_line)
            else:
                args = self.command_line.split(" ")
                self.set_arg("interval", args[0])
                if len(args) > 1:
                    self.set_arg("jitter", int(args[1]))
                else:
                    self.set_arg("jitter", 0)

    async def parse_dictionary(self, dictionary_arguments):
        self.load_args_from_dictionary(dictionary_arguments)


async def update_sleep_info(task: PTTaskCompletionFunctionMessage):
    await SendMythicRPCCallbackUpdate(
        MythicRPCCallbackUpdateMessage(
            CallbackID=task.TaskData.Callback.ID,
            SleepInfo=task.TaskData.Task.DisplayParams,
        )
    )


# Callback function to display output to Mythic
async def formulate_output(
    task: PTTaskCompletionFunctionMessage,
) -> PTTaskCompletionFunctionMessageResponse:
    # Get the interval and jitter from the task information
    params = json.loads(task.TaskData.Task.Params)
    interval = params["interval"]
    jitter = params["jitter"]

    # Format the output message
    output = f"Set sleep interval to {interval} seconds with a jitter of {jitter}%."

    await SendMythicRPCResponseCreate(
        MythicRPCResponseCreateMessage(
            TaskID=task.TaskData.Task.ID,
            Response=output.encode(),
        )
    )


async def post_run_actions(
    task: PTTaskCompletionFunctionMessage,
) -> PTTaskCompletionFunctionMessageResponse:
    try:
        await formulate_output(task)
        await update_sleep_info(task)
    except Exception as e:
        output = "".join(traceback.format_exception(e))
        output = f"Error during post processing:\n{output}"

        await SendMythicRPCResponseCreate(
            MythicRPCResponseCreateMessage(TaskID=task.TaskData.Task.ID, Response=output.encode())
        )

        return PTTaskCompletionFunctionMessageResponse(
            TaskID=task.TaskData.Task.ID,
            ParentTaskId=0,
            Completed=True,
            Success=False,
            Error=output,
            Stderr=output,
        )

    return PTTaskCompletionFunctionMessageResponse(
        TaskID=task.TaskData.Task.ID,
        ParentTaskId=0,
        Completed=True,
        Success=True,
    )


class SleepCommand(CommandBase):
    cmd = "sleep"
    needs_admin = False
    help_cmd = "sleep [number][suffix] [jitter]"
    description = "Change the agent's sleep interval. Suffix can either be [s, m, h]"
    version = 2
    author = "@M_alphaaa"
    argument_class = SleepArguments
    completion_functions = {"post_run_actions": post_run_actions}
    attackmapping = ["T1029"]
    attributes = CommandAttributes(
        supported_os=[SupportedOS.Linux, SupportedOS.Windows],
        builtin=True,
    )

    async def create_tasking(self, task: MythicTask) -> MythicTask:
        interval = task.args.get_arg("interval")
        jitter = task.args.get_arg("jitter")

        try:
            # Try to directly convert this interval to seconds
            try:
                real_interval = int(interval)
                units = "s"
            except ValueError:
                # Set the units to the last charater in the interval
                units = interval[-1]

                # Strip out the unit suffix of the interval
                interval = interval[:-1]

                if units == "s":  # Units are seconds
                    conversion_factor = 1
                elif units == "m":  # Units are minutes
                    conversion_factor = 60
                elif units == "h":  # Units are hours
                    conversion_factor = 3600
                else:
                    raise Exception("Invalid interval suffix [s, m, h]")

                # Convert the inputted interval to seconds using the conversion factor
                try:
                    real_interval = int(interval) * conversion_factor
                except ValueError:
                    raise Exception("Invalid sleep interval")

            # Check that the interval is not negative
            if real_interval < 0:
                raise Exception("Interval cannot be negative")

            # Make sure the jitter is not negative
            if jitter is not None and jitter < 0:
                raise Exception("Jitter cannot be negative")

            task.completed_callback_function = "post_run_actions"
        except Exception as e:
            output = "".join(traceback.format_exception(e))
            output = f"Error during command invocation:\n{output}"
            task.set_status(MythicStatus.Error)
            task.set_stderr(output)

        # Set the new interval
        task.args.remove_arg("interval")
        task.args.add_arg("interval", real_interval, type=ParameterType.Number)
        task.display_params = f"interval = {interval}{units}, jitter = {jitter}%"
        return task
